File: backend/src/main/java/com/devops/backend/service/CostMonitoringService.java
package com.devops.backend.service;

import com.devops.backend.dto.monitoring.CostTimelinePoint;
import com.devops.backend.dto.monitoring.LiveCostRow;
import com.devops.backend.dto.monitoring.ProjectCostBreakdown;

import java.time.Instant;
import java.util.List;

/**
 * DevOps-only cost tracking. Uses the latest Azure retail prices and the
 * current running state of each cloud service to produce real-time USD
 * figures, including shared-resource splits.
 */
public interface CostMonitoringService {

    /**
     * Re-apply the latest cached Azure prices to every cloud service and
     * every cluster node pool. Called after a pricing refresh.
     *
     * @return number of rows whose hourlyRateUsd was updated.
     */
    int applyLatestPricesToProjects();

    /** Capture a new cost snapshot for every currently running service. */
    int tickLiveCosts();

    /** Start a running cycle for a specific cloud service on a project/env. */
    void startCycle(String projectId, String environment, String cloudServiceId, String actor);

    /** Stop the current running cycle (if any) for a cloud service. */
    void stopCycle(String projectId, String environment, String cloudServiceId, String actor);

    /** Live breakdown across all projects for DevOps. */
    List<LiveCostRow> getLiveCosts();

    /** Per-project breakdown — totals plus per-service rows. */
    ProjectCostBreakdown getProjectBreakdown(String projectId, String environment);

    /** Historical cost timeline for a project (or service) over a window. */
    List<CostTimelinePoint> getCostTimeline(String projectId, String environment, String cloudServiceId,
                                            Instant from, Instant to);
}

File: frontend/src/services/costMonitoringService.js
import { apiRequest } from "./apiClient";

const q = (obj) => {
    const p = new URLSearchParams();
    Object.entries(obj || {}).forEach(([k, v]) => {
        if (v !== undefined && v !== null && v !== "") p.set(k, String(v));
    });
    const s = p.toString();
    return s ? `?${s}` : "";
};

/** DevOps + Admin: every cloud service across every project — live. */
export const getLiveCosts = () => apiRequest("/cost-monitoring/live");

export const getProjectBreakdown = ({ projectId, environment }) =>
    apiRequest(`/cost-monitoring/project${q({ projectId, environment })}`);

export const getCostTimeline = ({ projectId, environment, cloudServiceId, from, to }) =>
    apiRequest(`/cost-monitoring/timeline${q({ projectId, environment, cloudServiceId, from, to })}`);

/** DevOps only: start / stop a running cycle for a service. */
export const setServiceCycle = ({ projectId, environment, cloudServiceId, action }) =>
    apiRequest("/cost-monitoring/cycle", {
        method: "POST",
        body: JSON.stringify({ projectId, environment, cloudServiceId, action }),
    });

export const forceTick = () => apiRequest("/cost-monitoring/tick", { method: "POST" });

/**
 * Capacity-based breakdown per environment — feeds the DevOps "Cost Management"
 * dashboard (totals per env, per-project share of node capacity, savings hints).
 */
export const getCapacityBreakdown = () =>
    apiRequest("/cost-monitoring/capacity-breakdown");

export const getCapacityBreakdownFor = (environmentId) =>
    apiRequest(`/cost-monitoring/capacity-breakdown/${encodeURIComponent(environmentId)}`);
